src/main.ts

```diff
@@ -131,24 +198,26 @@ export default class ExecuteCodePlugin extends Plugin {
 	 * @param srcCode The code in the code block.
 	 * @param button The button element to which the listener is added.
 	 * @param out The {@link Outputter} object that is used to display the output of the code.
+	 * @param file The file that the code originates in
 	 */
-	private addListenerToButton(language: string, srcCode: string, button: HTMLButtonElement, out: Outputter) {
-		if (language.contains("language-js") || language.contains("language-javascript")) {
+	private addListenerToButton(language: LanguageId, srcCode: string, button: HTMLButtonElement, out: Outputter, file: string) {
+		if (language == "js") {
```

Comment:
Could use a switch-case statement. Would be good to replace weak equality with strict equality (===) everywhere.

Reply:
I agree regarding the strict equality (committed!), but I'm respectfully not sure about the switch/case. It would definitely be better practice, but I think that the `if`/`else` tree is trending towards a future where all of the branches are equivalent:

```ts
button.addEventListener("click", async () => {
	button.className = runButtonDisabledClass;
	const transformedCode = await new CodeInjector(this.app, this.settings, language).injectCode(srcCode);
//Differences in code (e.g. shell/no shell, prolog, CPP's `std=` argument) could
//be handled by executors, splitting code up and keeping better style
	this.executors.getExecutorFor(file, language) //these properties of `settings` will cause compile errors, but they'll be fine at runtime. Would use a cast in real code, but this is just a demo
	.run(transformedCode, out, this.settings[language + "Path"], this.settings[language + "Args"], this.settings[language + "FileExtension"])
	.then(() => {
		button.className = runButtonClass;
		out.closeInput();
	});
});
```

I'm not sure if such a large change (and one which would make it difficult for others to resolve merge conflicts) makes sense right now, when i could refactor more, in one patch of changes, once this PR merges. I'm definitely open to hearing your opinion on it though! Thank you for your review!

Reply:
I think you're right about removing the branching, but I also think executors for each language could be a bit too much boilerplate if the language doesn't support notebook mode, considering the different branches are mostly just for running a couple transformations on the code and command args. Maybe we could just have a map with functions for each language that just do this small bit of logic, or some other simpler alternative to executors? In any case, I think this would be best in a new PR.

Reply:
I think for the most languages without Notebook mode we can use the same generic Executor and don't have to implement a new one for each.

I think it's time to gt rid of the large if-branching. It started out with 2 or 3 different languages but the if branch is getting getting out of hand lately.

Reply:
I also have a strong dislike for boilerplate, but I'm worried that a map of functions could grow to be as unmaintainable as the current branching. I think we need some way to only specify certain exceptions, and have sensible defaults for all other languages.

The generic Executor needs the `shell` parameter, but that could be solved by a variation on Milan's map idea: with a map of booleans instead of functions, we can also default to `false` so that it's less verbose.

Other than `shell:true`/`shell:false`, the only inconsistencies are file extensions (should maybe be a setting for all languages), magic functions, inline plots, `out.clear()` being called in Prolog and CPP, and `-std` for CPP. 

Is there a reason why other languages *don't* call `out.clear()`? Can we handle magic functions and inline plots in the code injector, or abstract their per-language branching away into `Magic.ts`? If we could handle both of these, then the only remaining issue is CPP, which *could* be put in the generic executor. The generic executor already has language-specific behaviour for CPP (as taken from the `runCode` function), so I don't think it would be as problematic (still a code smell, but tolerable for now?)

Reply:
> Other than shell:true/shell:false, the only inconsistencies are file extensions 
I needed to set shell to true for all languages as quick fix for problems with the path (#77 #81)

> Is there a reason why other languages don't call out.clear()?
I think this call is unnecessary in C++ too but I am not sure.